react.md

```diff
@@ -0,0 +1,242 @@
+# Trunk Club React Styleguide
+
+This guide assumes that you are writing in the coffeescript-flavor of [JSX](https://github.com/jsdf/coffee-react-transform). It also assumes that you have a cursory understanding of what JSX is and why React uses it. See the [docs](http://facebook.github.io/react/docs/jsx-in-depth.html) for more details.
+
+## Structure of React Components
+You should structure your React Components like this to ensure consistency. Whether you're using ES6 or Coffeescript or PoopyScript, it should be ordered like this
+
+1. Attributes
+  1. If you are using a mixin that requires you to specify an attribute on the Component's prototype (and not on the instance itself), do that here as well
+
+  ```Coffeescript
+  ComponentName = React.createClass
+    displayName: 'ComponentName'
+    mixins: []
+    statics: {}
+    propTypes: {}
+    listen: {} # Custom attribute needed by a mixin
+  ```
+
+2. React lifecycle hooks
+
+  ```Coffeescript
+  getInitialState: -> {}
+  componentWillMount: ->
+  componentDidMount: ->
+  componentWillReceiveProps: (nextProps) ->
+  shouldComponentUpdate: (nextProps, nextState) -> true
+  componentWillUpdate: (nextProps, nextState) ->
+  componentDidUpdate: (prevProps, prevState) ->
+  ```
+
+3. Event handlers
+
+  ```Coffeescript
+  onClickBtn: (evt) -> # on(EventType)(SourceItem)
+                       # the React docs will use `handle(EventType)`.
+                       # Same thinking, different naming
+  ```
+
+4. Helper methods for making chunks of UI
+
+  ```CoffeeScript
+  makeListOfItems: ->
+    @state.items.map (item, index) -> <li key-{index}>{item}</li>
+  ```
+
+5. `render()`
+
+  ```CoffeeScript
+  render: ->
+    <span>{@makeListOfItems()}</span>
+  ```
+
+## Code style
+1. Make sure that you are returning a single root element inside of `render()`
+
+  ```CoffeeScript
+  ###
+  BAD EXAMPLE:
+  -----------
+  This example will function as expected due to the way that coffeescript compiles.
+  
+  The last token in a scope will be returned, which means that the last compiled JSX
+  node (the `<div>`) will be returned, and the first one will not.
+  ###
+  render: ->
+    <div>Here is some content</div>
+    <div>
+      <p>Here is some more</p>
+    </div>
+
+  # GOOD EXAMPLE:
+  render: ->
+    <section>
+      <header>
+        <nav>Here is the navigation</nav>
+      </header>
+      <div>
+        <p>Here is the first text on the page!</p>
+      </div>
+    </section>
+  ```
+
+2. If you have more than 3 props on a React component, move each prop declaration to a new line
```

Comment:
In a brief googley search, it looks like most other people have a "when it gets untidy OR 4 attributes, whichever comes first" rule, which is alright by me. There's cases like this:

```CoffeeScript
<img
  src="google.com/img.png"
  key={1}
  title="x"
  alt="y"
/>
```

Where putting each prop on its own line is more trouble than it's worth. It's hazy and contextual, since everyone has a different definition of "it's getting messy" or "this is hard to read".

How about: _"If the line has more than 80 characters OR has more than 5 props defined on it, whichever comes first"_